mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -246,4 +246,135 @@ class BlockMatrix(
+   */
+  def repartition(): BlockMatrix = {
+    repartition(rowsPerBlock, colsPerBlock)
+  }
+
+  /**
+   * Forms [[MatrixBlock]]s using the provided dimensions `newRowsPerBlock`x`newColsPerBlock`
+   * Even if all [[MatrixBlock]]s (except the ones on the right and bottom edge)
+   * have dimensions `newRowsPerBlock`x`newColsPerBlock`, a new [[BlockMatrix]] will be returned.
+   * The resulting number of partitions may be different than the current number of partitions.
+   *
+   * @param newRowsPerBlock Number of rows that make up each MatrixBlock.
+   * @param newColsPerBlock Number of columns that make up each MatrixBlock.
+   * @return The repartitioned BlockMatrix
+   */
+  def repartition(newRowsPerBlock: Int, newColsPerBlock: Int): BlockMatrix = {
+    repartition(newRowsPerBlock, newColsPerBlock, blocks.partitions.length)
+  }
+
+  /**
+   * Forms [[MatrixBlock]]s using the provided dimensions `newRowsPerBlock`x`newColsPerBlock`
+   * Even if all [[MatrixBlock]]s (except the ones on the right and bottom edge)
+   * have dimensions `newRowsPerBlock`x`newColsPerBlock`, a new [[BlockMatrix]] will be returned.
+   * The resulting number of partitions may be different than `suggestedNumPartitions`. Assumes
+   * that the offsets of each block is
+   * (`blockRowIndex`x`rowsPerBlock`, `blockColIndex`x`colsPerBlock`).
+   *
+   * If there are any [[MatrixBlock]]s with dimensions greater than `rowsPerBlock`x`colsPerBlock`,
+   * the data outside this scope will be added on to the data of the neighboring [[MatrixBlock]]s.
+   *
+   * @param newRowsPerBlock Number of rows that make up each block.
+   * @param newColsPerBlock Number of columns that make up each block.
+   * @param suggestedNumPartitions Number of partitions to partition the underlying RDD in. The
+   *                               final number of partitions may not equal this number.
+   * @return The repartitioned BlockMatrix
+   */
+  def repartition(
+      newRowsPerBlock: Int,
+      newColsPerBlock: Int,
+      suggestedNumPartitions: Int): BlockMatrix = {
+    require(newRowsPerBlock > 0,
+      s"newRowsPerBlock must be greater than 0. newRowsPerBlock: $newRowsPerBlock")
+    require(newColsPerBlock > 0,
+      s"newColsPerBlock must be greater than 0. newColsPerBlock: $newColsPerBlock")
+    require(suggestedNumPartitions > 0, s"suggestedNumPartitions must be greater than 0. " +
+      s"suggestedNumPartitions: $suggestedNumPartitions")
+    val m = numRows()
+    val n = numCols()
+    val newNumRowBlocks = math.ceil(m * 1.0 / newRowsPerBlock).toInt
+    val newNumColBlocks = math.ceil(n * 1.0 / newColsPerBlock).toInt
+    val slicedBlocks = blocks.flatMap { case ((blockRowIndex, blockColIndex), mat) =>
+      val rowStartOffset = blockRowIndex * rowsPerBlock
+      // val rowEndOffset = rowStartOffset + mat.numRows
+      val colStartOffset = blockColIndex * colsPerBlock
+      // val colEndOffset = colStartOffset + mat.numCols
+
+      /*
```

Comment:
I left it there to discuss which version would be more efficient. The commented version is super ugly, but should be more efficient.
We just never got to discussing it :/